bson/primitive/decimal_test.go

```diff
@@ -144,3 +148,61 @@ func TestParseDecimal128(t *testing.T) {
 		}
 	}
 }
+
+func TestDecimal128_JSON(t *testing.T) {
+	t.Run("roundTrip", func(t *testing.T) {
+		decimal := NewDecimal128(0x3040000000000000, 12345)
+		bytes, err := json.Marshal(decimal)
+		assert.Nil(t, err, "json.Marshal error: %v", err)
+		got := NewDecimal128(0, 0)
+		err = json.Unmarshal([]byte(bytes), &got)
+		assert.Nil(t, err, "json.Unmarshal error: %v", err)
+		assert.Equal(t, decimal.h, got.h, "expected h: %v got: %v", decimal.h, got.h)
+		assert.Equal(t, decimal.l, got.l, "expected l: %v got: %v", decimal.l, got.l)
+	})
+	t.Run("unmarshal extendedJSON", func(t *testing.T) {
+		want := NewDecimal128(0x3040000000000000, 12345)
+		extJSON := fmt.Sprintf(`{"$numberDecimal": %q}`, want.String())
+
+		got := NewDecimal128(0, 0)
+		err := json.Unmarshal([]byte(extJSON), &got)
+		assert.Nil(t, err, "json.Unmarshal error: %v", err)
+		assert.Equal(t, want.h, got.h, "expected h: %v got: %v", want.h, got.h)
+		assert.Equal(t, want.l, got.l, "expected l: %v got: %v", want.l, got.l)
+	})
+	t.Run("unmarshal null", func(t *testing.T) {
+		want := NewDecimal128(0, 0)
+		extJSON := `{"$numberDecimal": null}`
```

Comment:
Is this the behavior we want? My suggestion for handling `null` was just to handle document fields like `{a: "null"}`, not `{a: {$numberDecimal: "null"}}`.The special handling for "null" comes from a push in the Go community to allow `null` values to be unmarshalled into non-pointer struct fields. This gains convenience when communicating across languages and API boundaries at the expense of type safety and roundtrip-ability. However, IMO, if we get as far as having a `$numberDecimal` field, we should expect it to be a proper extjson decimal 128 string.

Reply:
Agreed. That's what I meant for this test to be, and it turns out that `{"$numberDecimal": null}` correctly fails to unmarshal